### PR TITLE
indexserver: use repo IDs instead of names

### PR DESCRIPTION
**Pull request by keegancsmith**
This moves the bulk of our interactions with Sourcegraph to be based on repository IDs instead of names. To do this required also updating our cleanup routines to be ID based.

Cleanup is based on the list of assigned repositories. We could keep the name based cleanup if we did it after we had resolved options. However, resolving options can fail so I opted to update the cleanup code to be ID based. This is where the bulk of the changes occurred. We set tombstones as part of cleanup, so that also switched to being ID based.

### CODE REVIEW

**Comment by keegancsmith** [comment]
@tsenart Marking this as ready for review. There is some minor work still on this branch, and I am currently testing it a bunch, especially interactions with Sourcegraph. I don't expect major changes so marking for review now. Will push any other changes as separate commits to this PR so they can be reviewed separately.

Remaining TODOs:

- [x] Regression on TSV logs, we now say ID which isn't as debuggable as name. The TSV logs are for us to debug, so need to fix before landing.
- [x] Need to think through corner case around "name swap" of a private and public repo. Also should manually test.
- [x] Should we introduce a tiny type for repo id? Will do for follow up if so.

**Review by tsenart — APPROVED** [review]

**Comment by keegancsmith** [comment]
Did a bunch of manual testing here, especially around privacy, name changes and ID changes. We straight away update indexes and at worse on the second run of the loop cleanup. I found this code quite useful to dynamically updating the ID in our fake:

```diff
diff --git a/cmd/zoekt-sourcegraph-indexserver/main.go b/cmd/zoekt-sourcegraph-indexserver/main.go
index 321185c..0cd219e 100644
--- a/cmd/zoekt-sourcegraph-indexserver/main.go
+++ b/cmd/zoekt-sourcegraph-indexserver/main.go
@@ -20,6 +20,7 @@ import (
 	"os/exec"
 	"os/signal"
 	"path/filepath"
+	"reflect"
 	"runtime"
 	"sort"
 	"strconv"
@@ -254,6 +255,21 @@ func (s *Server) Run() {
 				continue
 			}

+			if b, err := os.ReadFile(filepath.Join(s.IndexDir, "id.txt")); err == nil {
+				fields := bytes.Fields(b)
+				ids := map[string]uint32{}
+				for i := 1; i < len(fields); i += 2 {
+					id, _ := strconv.Atoi(string(fields[i]))
+					ids[string(fields[i-1])] = uint32(id)
+				}
+				fakeIDsMu.Lock()
+				if !reflect.DeepEqual(ids, fakeIDs) {
+					log.Printf("fake IDs changed: %v", ids)
+					fakeIDs = ids
+				}
+				fakeIDsMu.Unlock()
+			}
+
 			repos, err := s.Sourcegraph.ListRepoIDs(context.Background(), listIndexed(s.IndexDir))
 			if err != nil {
 				log.Println(err)
diff --git a/cmd/zoekt-sourcegraph-indexserver/sg.go b/cmd/zoekt-sourcegraph-indexserver/sg.go
index b644f59..d92bfec 100644
--- a/cmd/zoekt-sourcegraph-indexserver/sg.go
+++ b/cmd/zoekt-sourcegraph-indexserver/sg.go
@@ -16,6 +16,7 @@ import (
 	"path"
 	"path/filepath"
 	"strconv"
+	"sync"

 	"github.com/google/zoekt"
 	retryablehttp "github.com/hashicorp/go-retryablehttp"
@@ -338,8 +339,18 @@ func (sf sourcegraphFake) visitRepos(visit func(name string)) error {
 	})
 }

+var fakeIDsMu sync.Mutex
+var fakeIDs = map[string]uint32{}
+
 // fakeID returns a deterministic ID based on name. Used for fakes and tests.
 func fakeID(name string) uint32 {
+	fakeIDsMu.Lock()
+	id := fakeIDs[name]
+	fakeIDsMu.Unlock()
+	if id > 0 {
+		return id
+	}
+
 	// magic at the end is to ensure we get a positive number when casting.
 	return uint32(crc32.ChecksumIEEE([]byte(name))%(1<<31-1) + 1)
 }
```
